### ISSUE TITLE
Follow-up of #8493 - JSON-API tries 2FA and fails for sure, $third_party misused?

### ISSUE THREAD

**Issue by Quix0r** [friendica/friendica#10389]
With regard to #8493, I opened another ticket. So here are my findings:

1. 2FA is enabled as I prefer it safer (than sorry).
2. API calls on `/api/account/verify_credential.json` end in an endless loop and no authentication takes place
3. Since there is no 2FA possible for JSON calls, this needs excluding there, else no auth will happen over API calls.

I then checked the source code and found out that `/include/api.php` sets `$third_party` to `true` which seem to cause the endless loop of authentication requests (you can try it out yourself, please have 2FA enabled for this test!).

The solution here I propose is that `$third_party` remains `false and `true` is being removed from `/include/api.php`'s invocation of `User::authenticate()` as the API is not really a 3rd-party application but an API.

**Comment by Quix0r** [comment]
I think the `$api` in `User::authenciate()` makes no sense as `$third_party` needs to be turned off so the misuse isn't happening. I guess more that `Authentication->setForUser()` needs to check if an API is invoked and then bypass the redirect to 2FA verification as JSON clients cannot do that.

**Edit:**
Okay, `Authentication->redirectForTwoFactorAuthentication()` does bypass API calls? That's now confusing? Why is there then `$third_party` set?

**Edit 2:**
If I remove that said flag from `/include/api.php`, I get a JSON back with successful authorization. Now only @Prosody doesn't want it?

**Edit3:**
Yes, this fix works then, I'm now logged into prosody with @Dino !
````
local have_async, async = nil;
-- local have_async, async = pcall(require, "util.async");
````
Hmm, cannot this last one be fixed differently? That Friendica can tell prosody's module to avoid async utility? That would be a more flexible solution than manual hacking `mod_auth_http_async` (while there is `mod_auth_http`, [too](https://modules.prosody.im/mod_auth_http.html)?)

**Comment by AlfredSK** [comment]
Really, I have no idea why you have those troubles. I have set this up w/o any issues and w/o any "hacking" for Prosody and Nextcloud. With 2FA and app tokens it works too.

**Comment by Quix0r** [comment]
Yes, strange. Looks like I attract bugs a lot... :-( This small change was needed in `/include/api.php` of Friendica:
````
diff --git a/include/api.php b/include/api.php
index 0b925aee0..0980abce6 100644
--- a/include/api.php
+++ b/include/api.php
@@ -257,7 +257,7 @@ function api_login(App $a, bool $do_login = true)
        if ($addon_auth['authenticated'] && !empty($addon_auth['user_record'])) {
                $record = $addon_auth['user_record'];
        } else {
-               $user_id = User::authenticate(trim($user), trim($password), true);
+               $user_id = User::authenticate(trim($user), trim($password));
                if ($user_id !== false) {
                        $record = DBA::selectFirst('user', [], ['uid' => $user_id]);
                }
````
And oh, I hope you don't mind that I added type-hints to a deprecated method `User::authenticate()`?

**Comment by Quix0r** [comment]
PR #10390 contains these changes now.

**Comment by Quix0r** [comment]
This change wouldn't be needed if proper documentation (e.g. in the actual view, visible to the user), that the user needs to setup an "app-specific password" for prosody (the user then has to use instead of the actual password. It wasn't clear enough and may not sound so obvious as regular people don't understand the "technical" details behind this and will use their normal passwords instead and wonder why they cannot login.

**Comment by MrPetovan** [comment]
There: https://f.haeder.net/help/Two-Factor-Authentication#Third-party+applications+and+API

**Comment by Quix0r** [comment]
Yes. somehow it wasn't obvious here. Maybe some JSON message that tells the user over the program (e.g. Dino) that normal passwords cannot be used with 2FA and that the user has to setup an application-specific password.

**Edit:** It does work flawless here (maybe I still need the change to `mod_auth_http_async.lua`).

**Comment by MrPetovan** [comment]
A specific error message sounds good.
